Make election behavior mutable to support dynamic behaviour transition

gr0vity-dev opened the pull request:
Problem:
- Elections are currently immutable and can't be transitioned to priority status
- Causing delays for optimistic elections rescheduled by priority scheduler as the next Vote broadcast is fixed to 15-second intervals. (by `network.vote_broadcast_interval`)

Changes:
- Modifies election behavior to be mutable, allowing elections (optimistic/hinted/...) to be transitioned to priority status
- Added ability to trigger immediate vote broadcasts on transition without waiting for the next 15-second round

Benefits:
- It would allow for more aggressive use of optimistic elections (to narrow the gap between checked and cemented during bootstrapping) 
  - currently configured to accounts with gap_threshold of 32 could be reduced to 4
- Maintain quick response times for live traffic by allowing election behavior changes.

pwojcikdev commented:
Can you include a test or two that ensures this works? So I imagine and election would be manually inserted via `active.insert ( ..., election_behavior::optimistic )` and then upgraded to priority.

gr0vity-dev commented:
I added a simple testcase to validate the behavior transitioning works.

Ideally I'd like to also make sure the `vote_broadcast_interval` is reset and a rep votes again immediately after transitioning.
My problem is to create a network with one rep that doesn't have enough weight to confirm an election.
It seems :
- we need to start from genesis and transfer some weight over to one rep
- by creating a  send + open block and confirm these.  

If that approach is right I can create another testcase that makes sure voting happens as intended.

pwojcikdev commented on nano/node/active_elections.cpp:
```diff
@@ -287,8 +287,9 @@ void nano::active_elections::cleanup_election (nano::unique_lock<nano::mutex> &
 	// Keep track of election count by election type
-	debug_assert (count_by_behavior[election->behavior ()] > 0);
-	count_by_behavior[election->behavior ()]--;
+	auto election_behavior = election->behavior_locked (lock_a);
```
This change seems wrong, I'm surprised it compiles. 1) it's passing `active_elections`s lock into the election, 2) behavioud_locked should be private function of the election, not available outside

gr0vity-dev replied:
You are right.  I'll revert the change, and look more closely at why TSAN was complaining.

pwojcikdev commented on nano/node/election.cpp:
```diff
@@ -776,8 +776,15 @@ std::vector<nano::vote_with_weight_info> nano::election::votes_with_weight () co
 
+nano::election_behavior nano::election::behavior_locked (nano::unique_lock<nano::mutex> & lock_a) const
+{
+	debug_assert (lock_a.owns_lock ());
```
Missing assert for` !mutex.try_lock ()`, it doesn't check that the lock passed specifically is locking this election mutex

pwojcikdev replied:
It doesn't use the lock anyway, so it's unnecessary, only the assert for `!mutex.try_lock ()` should be here.

pwojcikdev commented on nano/node/active_elections.cpp:
```diff
@@ -435,6 +436,19 @@ nano::election_insertion_result nano::active_elections::insert (std::shared_ptr<
+		{
+			count_by_behavior[previous_behavior]--;
+			count_by_behavior[election_behavior_a]++;
+			result.election->transition_priority ();
```
Is this guaranteed to always succeed? This probably works but is unnecessary leaving space for subtle bugs. Should be something like:
```
bool transitioned = result.election->transition_priority ();
if (transitioned)
{
// success, update counts, stats, log
}
else
{
// failed for some reason, inc failed stats
}
```

pwojcikdev commented:
Regarding testing the voting behaviour, I don't think having two nodes is necessary. Each election should keep counters for number of vote broadcasts and confirmation requests and those can be queried directly. There is still a need to setup a rep key for node with weight higher than primary rep threshold but not enough to confirm blocks on its own, but that's rather simple.

gr0vity-dev commented:
> Each election should keep counters for number of vote broadcasts and confirmation requests
My issue is
-  by default the testcase node is not voting and doesn't have a rep wallet in it. 
- So it doesn't broadcast a vote.
https://github.com/nanocurrency/nano-node/blob/a4614215d030cfcb3b39799fd4ded2f72bbb3f63/nano/node/election.cpp#L429
I wanted to add a rep wallet, but
-  if I use the genesis key, blocks are confirmed by its single vote 
- So the optimistic election is not activated anymore...

pwojcikdev commented:
>if I use the genesis key, blocks are confirmed by its single vote
So the optimistic election is not activated anymore...


You should be able to setup a representative with eg. 111K nano weight which will be generating votes but not be able to confirm elections. Am I missing something in this setup?

gr0vity-dev commented:
Yes exactly. My initial question was how to best do it. I didn't see see a helper method to set this up.
The best I found was :
 - Start from genesis and transfer some weight over to one rep by creating a send + open block and force confirm these blocks.

This requires quite a bit of code. If that's the right way, I'll go for it. Just wanted to make sure this is the right approach.

pwojcikdev commented:
There is `nano::test::setup_rep ( ... )` if that's helpful.

Commit:
Make election behaviour mutable

Commit:
Add transition_priority method to election class

Commit:
Make use of transition_priority() in active_elections

Commit:
Add testcase election_scheduler.transition_optimistic_to_priority

Commit:
extend election status with vote_broadcast_count and adjust testcase

Commit:
simplify election_status initialization

Replace inline initializations with the new constructor syntax while maintaining the same behavior.
